gpbft/chain_test.go

```diff
@@ -0,0 +1,138 @@
+package gpbft_test
+
+import (
+	"bytes"
+	"encoding/binary"
+	"testing"
+
+	"github.com/filecoin-project/go-f3/gpbft"
+	"github.com/stretchr/testify/require"
+)
+
+func TestTipSet(t *testing.T) {
+	t.Parallel()
+	tests := []struct {
+		name       string
+		subject    gpbft.TipSet
+		wantZero   bool
+		wantString string
+	}{
+		{
+			name:       "zero-value struct is zero",
+			wantZero:   true,
+			wantString: "@0",
+		},
+		{
+			name:       "ZeroTipSet is zero",
+			subject:    gpbft.ZeroTipSet(),
+			wantZero:   true,
+			wantString: "@0",
+		},
+		{
+			name:       "NewTipSet with zero values is zero",
+			subject:    gpbft.NewTipSet(0, gpbft.NewTipSetID(nil)),
+			wantZero:   true,
+			wantString: "@0",
+		},
+		{
+			name:       "Non-zero is not zero",
+			subject:    gpbft.NewTipSet(1413, gpbft.NewTipSetID([]byte("fish"))),
+			wantString: "fish@1413",
+		},
+		{
+			name:       "negative epoch is accepted",
+			subject:    gpbft.NewTipSet(-1413, gpbft.NewTipSetID([]byte("fish"))),
+			wantString: "fish@-1413",
+		},
+	}
+	for _, test := range tests {
+		test := test
+		t.Run(test.name, func(t *testing.T) {
+			require.Equal(t, test.wantZero, test.subject.IsZero())
+			require.Equal(t, test.wantString, test.subject.String())
+			requireTipSetMarshaledForSigning(t, test.subject)
+		})
+	}
+}
+
+func requireTipSetMarshaledForSigning(t *testing.T, subject gpbft.TipSet) {
+	t.Helper()
+	var gotSigningMarshal bytes.Buffer
+	subject.MarshalForSigning(&gotSigningMarshal)
+	wantPrefix := binary.BigEndian.AppendUint64(nil, uint64(subject.Epoch))
+	wantSuffix := subject.CID.Bytes()
+	require.Equal(t, len(wantPrefix)+len(wantSuffix), gotSigningMarshal.Len())
+	require.True(t, bytes.HasPrefix(gotSigningMarshal.Bytes(), wantPrefix))
+	require.True(t, bytes.HasSuffix(gotSigningMarshal.Bytes(), wantSuffix))
+}
+
+func TestECChain(t *testing.T) {
+	t.Parallel()
+
+	zeroTipSet := gpbft.ZeroTipSet()
+	t.Run("zero-value is zero", func(t *testing.T) {
+		var subject gpbft.ECChain
+		require.True(t, subject.IsZero())
+		require.False(t, subject.HasBase(zeroTipSet))
+		require.False(t, subject.HasPrefix(subject))
+		require.False(t, subject.HasTipset(zeroTipSet))
+		require.False(t, subject.SameBase(subject))
+		require.True(t, subject.Eq(subject))
+		require.True(t, subject.Eq(*new(gpbft.ECChain)))
+		require.Nil(t, subject.Suffix())
+		require.Panics(t, func() { subject.Prefix(0) })
+		require.Panics(t, func() { subject.Base() })
+		require.Panics(t, func() { subject.Head() })
+		require.Equal(t, zeroTipSet, subject.HeadOrZero())
+	})
+	t.Run("NewChain with zero tipset is not zero", func(t *testing.T) {
```

Comment:
I think it should be invariant that there is no zero tipset in a chain. Could you please address https://github.com/filecoin-project/go-f3/issues/128 and then change this into testing those invariants?

If you'd like to defer that to a follow-up PR, I suggest removing this test and replacing it with a link to #128.

Reply:
Thank you for pointing that out.
I left a TODO in this PR, and will address #128 along with tests in a follow-up PR shortly.

Reply:
Resolved by https://github.com/filecoin-project/go-f3/pull/138